### PR TITLE
#560 Improve dataloaders

### PR DESCRIPTION
**Pull request by Techno-Tut**
# Enchanced DataLoader

Added Custom caching capability allowing the cache to be completely modifiable as per the official data loader API specifications.
Added ability to set custom hashing function for data loader, enabling overriding the default behaviour, as per data loader specifications.

### custom caching
```python
from typing import List, Union, Any, Optional

import strawberry
from strawberry.types import Info
from strawberry.asgi import GraphQL
from strawberry.dataloader import DataLoader, AbstractCache

from starlette.requests import Request
from starlette.websockets import WebSocket
from starlette.responses import Response


class UserCache(AbstractCache):
    def __init__(self):
        self.cache = {}

    def get(self, key: Any) -> Union[Any, None]:
        return self.cache.get(key)   #fetch data from persistent cache

    def set(self, key: Any, value: Any) -> None:
        self.cache[key] = value   #store data in the cache

    def delete(self, key: Any) -> None:
        del self.cache[key]   #delete key from the cache

    def clear(self) -> None:
        self.cache.clear()  #clear the cache


@strawberry.type
class User:
    id: strawberry.ID
    name: str


async def load_users(keys) -> List[User]:
    return [User(id=key, name="Jane Doe") for key in keys]


class MyGraphQL(GraphQL):
    async def get_context(self, request: Union[Request, WebSocket], response: Optional[Response]) -> Any:
        return {
            "user_loader": DataLoader(
                load_fn=load_users,
                cache_map=UserCache()
            )
        }

@strawberry.type
class Query:
    @strawberry.field
    async def get_user(self, info: Info, id: strawberry.ID) -> User:
        return await info.context["user_loader"].load(id)


schema = strawberry.Schema(query=Query)
app = MyGraphQL(schema, graphiql=True)
```
code change examples
### Custom Hashing function 

```python
from typing import List, Union
from strawberry.dataloader import DataLoader

class User:
    def __init__(self, custom_id: int, name: str):
        self.id: int = custom_id
        self.name: str = name

async def loader_fn(keys):
    return keys

def custom_cache_key(key):
    return key.id

loader = DataLoader(load_fn=loader_fn, cache_key_fn=custom_cache_key)
data1 = await loader.load(User(1, "Nick"))
data2 = await loader.load(User(1, "Nick"))
assert data1 == data2 #returns true
```

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Implemented Custom Caching for dataloaders using `AbstractCache`
Added custom hashing function to dataloader `cache_key_fn`
Examples are in the documentation for `dataloader.md`


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @techno_tut for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Comment by Techno-Tut** [comment]
@patrick91 can i get a review here, I came for node side and I miss the data loader feature. I don't want to install a mew library just for dataloaders

File: docs/guides/dataloaders.md [hunk 1]
```diff
@@ -14,6 +14,8 @@ DataLoaders provide an async API, so they only work in async context
 
 </Note>
 
+Refer the official dataloaders [specification](https://github.com/graphql/dataloader) for advance guide on dataloaders.
+
```
Comment (patrick91) [inline docs/guides/dataloaders.md:18]:
```suggestion
Refer the official dataloaders [specification](https://github.com/graphql/dataloader) for an advanced guide on dataloaders.

```

File: strawberry/dataloader.py [hunk 2]
```diff
@@ -45,10 +45,50 @@ def __len__(self) -> int:
...
+    @abstractmethod
+    def delete(self, key: Any) -> None:
+        pass
+
+    @abstractmethod
+    def clear(self) -> None:
+        pass
+
+
+class DefaultCache(AbstractCache):
```
Comment (patrick91) [inline strawberry/dataloader.py:66]: we need to extend from generic here too 😊

File: strawberry/dataloader.py [hunk 3]
```diff
@@ -58,6 +98,8 @@ def __init__(
         max_batch_size: Optional[int] = None,
         cache: bool = True,
         loop: AbstractEventLoop = None,
+        cache_map: Optional[AbstractCache] = None,
+        cache_key_fn: Optional[Callable[[K], Any]] = None,
```
Comment (patrick91) [inline strawberry/dataloader.py:102]: shouldn't we use T instead of any here?

**Review by patrick91 — COMMENTED** [review]
Nice work so far, I've added some initial comments 😊

File: strawberry/dataloader.py (hunk 3 above)
Comment (Techno-Tut) [reply strawberry/dataloader.py]: I agree on this, will make it more predictable.

**Comment by Techno-Tut** [comment]
Hey @patrick91,  I have updated the PR, it now has more accurate type definitions, and the mypy test should now be back to normal. can you review it once you get time : )

**Comment by Techno-Tut** [comment]
Hey @patrick91, once you get time will you be able to run the workflow? I wana know the codecov report after fixing the mypy tests

**Comment by Techno-Tut** [comment]
raising a fresh one

**Pull request closed without merging**